src/ws-handler.js

```diff
@@ -0,0 +1,178 @@
+const JSONStream = require('JSONStream');
+const { pipeline } = require('stream');
+const url = require('url');
+const WebSocket = require('ws');
+
+const ClientSink = require('./ClientSink');
+const DemuxSink = require('./demux');
+const logger = require('./logging');
+const PromCollector = require('./metrics/PromCollector');
+const { getStatsFormat } = require('./utils/stats-detection');
+
+/**
+ *
+ */
+class WsHandler {
+
+    /**
+     *
+     */
+    constructor({ tempPath, reconnectTimout, sequenceNumberSendingInterval, dumpPersister }) {
+        this.sessionIdTimeouts = {};
+        this.tempPath = tempPath;
+        this.reconnectTimout = reconnectTimout;
+        this.sequenceNumberSendingInterval = sequenceNumberSendingInterval;
+        this.dumpPersister = dumpPersister;
+    }
+
+    /**
+     *
+     * @param {*} wsServer
+     */
+    setupWebSocketsServer(wsServer) {
+        const wss = new WebSocket.Server({ server: wsServer });
+
+        wss.on('connection', this._handle.bind(this));
+
+        return wss;
+    }
+
+    /**
+     * Main handler for web socket connections.
+     * Messages are sent through a node stream which saves them to a dump file.
+     * After the websocket is closed the session is considered as terminated and the associated dump
+     * is queued up for feature extraction through the {@code WorkerPool} implementation.
+     *
+     * @param {*} client
+     * @param {*} upgradeReq
+     */
+    _handle(client, upgradeReq) {
+        PromCollector.connected.inc();
+        logger.info('[App] Websocket connection handler');
+
+        // the url the client is coming from
+        const referer = upgradeReq.headers.origin + upgradeReq.url;
+        const ua = upgradeReq.headers['user-agent'];
+        const queryObject = url.parse(referer, true).query;
+        const statsSessionId = queryObject?.statsSessionId;
+
+        const connectionInfo = this._createConnectionInfo(upgradeReq, referer, ua, client);
+        const demuxSink = this._createDemuxSink(connectionInfo);
+        const clientSink = this._createClientSink(statsSessionId, demuxSink, client);
+
+        this._clearConnectionTimeout(statsSessionId);
+        clientSink.sendLastSequenceNumber(client, statsSessionId);
+
+        demuxSink.on('close-sink', ({ id, meta }) => {
+            logger.info('[App] Websocket disconnected waiting for processing the data %s', id);
```

Comment:
things to consider here:
1. What happens if the client reconnects but the timeout has expired, so the dump is no longer here, how will the client react when it receives a seq nr of 0 but it probably expects a larger sequence number.
A fast solution would be to just create a new dump with that statsessionid, fail to process it (bcs it would lack context) and just upload it to s3 with statssessionid + timestamp as the name so we don't loose the data.
2. What happens if the client reconnects and the websocket has not yet closed on the server side (it can happen), resulting in a demux already running for that statsessionid. We probably need to somehow forcefully close the current statsessionid demux before sending the lastsequencenr.